DecentralML - Milestone 1 Delivery

livetreetech opened the pull request:
# Milestone Delivery Checklist

- [x] The [milestone-delivery-template.md](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/deliveries/milestone-delivery-template.md) has been copied and updated.
- [x] This pull request is being made by the same account as the accepted application.
- [x] I have disclosed any and all sources of reused code in the submitted repositories and have done my due diligence to meet its license requirements.
- [x] In case of acceptance, an invoice must be submitted and the payment will be transferred to the Polkadot/fiat account provided in the application.
- [x] The delivery is according to the [Guidelines for Milestone Deliverables](https://grants.web3.foundation/docs/Support%20Docs/milestone-deliverables-guidelines).

Link to the application pull request: https://github.com/w3f/Grants-Program/pull/1079

takahser commented on deliveries/decentralml_milestone_1.md:
```diff
@@ -0,0 +1,36 @@
+| 0c. | Testing and Testing Guide | [LINK](https://github.com/livetreetech/DecentralML/blob/main/README.md) | | 
+| 0d. | Docker | [LINK](https://github.com/livetreetech/DecentralML/tree/main/docker) | | 
+| 1. | Data Management Implementation | [LINK](https://github.com/livetreetech/DecentralML/blob/main/README.md) |  See the DecentralML Rust pallet [LINK](https://github.com/livetreetech/DecentralML/blob/main/substrate-node-decentralml/pallets/decentralml/src/lib.rs) The calling Python client [LINK](https://github.com/livetreetech/DecentralML/tree/main/substrate-client-decentralml) The DecentralML ...
+| 2. | Federated Learning Consensus| [LINK](https://github.com/livetreetech/DecentralML/blob/main/substrate-node-decentralml/pallets/decentralml/src/lib.rs) | There is the deposit and financial administrative functions while creating the task i.e. depositing Currency; at the moment we are not using [ORML Currency](https://marketplace.substrate.io/pallets/orml-currencies any suggestions welcome he ...
```
```suggestion
| 2. | Federated Learning Consensus| [LINK](https://github.com/livetreetech/DecentralML/blob/main/substrate-node-decentralml/pallets/decentralml/src/lib.rs) | There is the deposit and financial administrative functions while creating the task i.e. depositing Currency; at the moment we are not using [ORML Currency](https://marketplace.substrate.io/pallets/orml-currencies) any suggestions welcome here) Note: "Data Contributors" can use TensorFlow's Federated Learning which for the moment is implemented through the parameter model_contributor_script_path inside the TaskInfo. We currently are exploring loading and executing this within a Docker based Python environment. NOTE: The defineDataContributors is the create_task "Origin" with the reward being deposited as part of the task creation. Validate_task_result will  deliver three validation strategies for task results: Auto which automatically transfers the worker the reward funds, Manual whereabouts the Creator needs to manually assigned or Custom which will faciliates more advanced validation methods. We have utilised Storage BoundedVectors to manage resources for optimal  performance as Task/user participation grows.
```

takahser reviewed (changes requested):
@livetreetech could you add the url to the source of the documentation as well (in addition to the deployed gitbook)?

livetreetech commented:
> @livetreetech could you add the url to the source of the documentation as well (in addition to the deployed gitbook)?

Not sure what you mean - does clicking this give you edit rights? https://app.gitbook.com/o/JGQQdHzEzlbR15Ezbn9M/s/Wt5ixztAjbKCHHNS4hEX/task-overview   if not please let me know how I get the edit link. Thank you.

takahser commented:
@livetreetech nope, it doesn't:
<img width="408" alt="image" src="https://github.com/w3f/Grant-Milestone-Delivery/assets/5393704/db3cde65-a83b-4903-9ddc-64bbbbcf3874">

I'm pretty sure you have a github repo that serves as a source for these docs. That's the URL that should be added to the delivery document.

livetreetech commented:
@takahser You can try this one https://app.gitbook.com/invite/JGQQdHzEzlbR15Ezbn9M/S6iumDpmx2X6wgd3gNb2
In any case it wont show you code documentation, you can run the cargo -h (see readme) to build all the code docs

takahser commented:
@livetreetech it didn't work, but I don't think it's necessary anyway.
<img width="389" alt="image" src="https://github.com/w3f/Grant-Milestone-Delivery/assets/5393704/cf8c8e6b-3042-4a44-8615-27c37c83cee3">

> In any case it wont show you code documentation, you can run the cargo -h (see readme) to build all the code docs

Ah true, according to the [embedded docs](https://github.com/livetreetech/DecentralML#embedded-docs) section in your README it seems like the docs are contained in your repo already, so that should be fine then.

semuelle commented:
Hey @livetreetech. Just a quick update: I haven't managed to go through all deliverables, but everything looks great so far. I hope to be able to wrap up the evaluation tomorrow.

livetreetech commented:
@semuelle thanks very much for the update

semuelle commented:
Hey @livetreetech. I am following the testing guide using Docker. I got to [this step](https://github.com/livetreetech/DecentralML/tree/485ec49606f68aa55d5d452bbafbe455caa246d1/docker#example), but then I get the following error running the command. Can you advise?

```sh
root@e9cbafd7a843:~/decentralml# python3 -m substrate-client-decentralml.create_task_ipfs
/usr/bin/python3: Error while finding module specification for 'substrate-client-decentralml.create_task_ipfs' (ModuleNotFoundError: No module named 'substrate-client-decentralml')
```

Unfortunately, the local install doesn't work either:

```sh
> poetry install
Traceback (most recent call last):
  File "/usr/bin/poetry", line 5, in <module>
    from poetry.console import main
  File "/usr/lib/python3/dist-packages/poetry/console/__init__.py", line 1, in <module>
    from .application import Application
  File "/usr/lib/python3/dist-packages/poetry/console/application.py", line 3, in <module>
    from cleo import Application as BaseApplication
ImportError: cannot import name 'Application' from 'cleo' (/home/dml/.local/lib/python3.10/site-packages/cleo/__init__.py)
```

semuelle commented:
pinging @livetreetech

livetreetech commented:
> Hey @livetreetech. I am following the testing guide using Docker. I got to [this step](https://github.com/livetreetech/DecentralML/tree/485ec49606f68aa55d5d452bbafbe455caa246d1/docker#example), but then I get the following error running the command. Can you advise?
> 
> 
> 
> ```sh
> 
> root@e9cbafd7a843:~/decentralml# python3 -m substrate-client-decentralml.create_task_ipfs
> 
> /usr/bin/python3: Error while finding module specification for 'substrate-client-decentralml.create_task_ipfs' (ModuleNotFoundError: No module named 'substrate-client-decentralml')
> 
> ```
> 
> 
> 
> Unfortunately, the local install doesn't work either:
> 
> 
> 
> ```sh
> 
> > poetry install
> 
> Traceback (most recent call last):
> 
>   File "/usr/bin/poetry", line 5, in <module>
> 
>     from poetry.console import main
> 
>   File "/usr/lib/python3/dist-packages/poetry/console/__init__.py", line 1, in <module>
> 
>     from .application import Application
> 
>   File "/usr/lib/python3/dist-packages/poetry/console/application.py", line 3, in <module>
> 
>     from cleo import Application as BaseApplication
> 
> ImportError: cannot import name 'Application' from 'cleo' (/home/dml/.local/lib/python3.10/site-packages/cleo/__init__.py)
> 
> ```

Sorry totally missed those, I'm on the rust side Mathias is on the python. I'll ask him tomorrow and come back to you sorry for the delay.

livetreetech commented:
> pinging @livetreetech

@semuelle 

Dr. Mathias Ciliberto — Today at 08:49
I have updated the documentation. The command is:
python3 -m decentralml.create_task_ipfs

semuelle commented:
Thanks for the update, @livetreetech. However, python can't seem to find a `create_task_ipfs` method. I'm not very familiar with Python, so if there is some step before/in between that I'm supposed to be doing, like an install or changing directories, please let me know. 

I am able to run `python3 -m decentralml.examples`. However, after selecting option 1, I don't see anything happening, even after waiting for several minutes. Is there debug output somewhere I could inspect?

```sh
node_decentralml    | 2023-12-19 10:17:09 💤 Idle (0 peers), best: #0 (0xb0b5…90c4), finalized #0 (0xb0b5…90c4), ⬇ 0 ⬆ 0    
node_decentralml    | 2023-12-19 10:17:14 💤 Idle (0 peers), best: #0 (0xb0b5…90c4), finalized #0 (0xb0b5…90c4), ⬇ 0 ⬆ 0    
client_decentralml  | root@e9cbafd7a843:/decentralml/substrate-client-decentralml# python3 -m decentralml.examples
Functionality to tests:
client_decentralml  | 1 - Create tasks
client_decentralml  | 2 - Assign tasks
client_decentralml  | 3 - Send task results
client_decentralml  | 4 - Exit
client_decentralml  | Select your example: 1
node_decentralml    | 2023-12-19 10:17:16 Accepting new connection 1/100
node_decentralml    | 2023-12-19 10:17:19 💤 Idle (0 peers), best: #0 (0xb0b5…90c4), finalized #0 (0xb0b5…90c4), ⬇ 0 ⬆ 0    
node_decentralml    | 2023-12-19 10:17:24 💤 Idle (0 peers), best: #0 (0xb0b5…90c4), finalized #0 (0xb0b5…90c4), ⬇ 0 ⬆ 0    
node_decentralml    | 2023-12-19 10:17:29 💤 Idle (0 peers), best: #0 (0xb0b5…90c4), finalized #0 (0xb0b5…90c4), ⬇ 0 ⬆ 0    
```

Locally, I still get the `ImportError: cannot import name 'Application' from 'cleo' (/home/dml/.local/lib/python3.10/site-packages/cleo/__init__.py)` error, so it'd be great if you could review either one and let me know how to proceed. Feel free to [reach out via Matrix](https://matrix.to/#/@sebastian:web3.foundation), might be faster.

semuelle reviewed (approved):
Hi @livetreetech. Thanks for the help, your milestone is hereby accepted. You can find my evaluation notes [here](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/evaluations/decentralML_1_semuelle.md).

meldien commented:
@livetreetech 
 Thank you for providing the invoice. 

Could you please add your full address and adding our billing information as well.

Web 3.0 Technologies Foundation
Baarerstrasse 14
6300 Zug

Many thank for adding the missing information and providing the invoice. 

Best,
Melanie

livetreetech commented:
> @livetreetech Thank you for providing the invoice.
> 
> Could you please add your full address and adding our billing information as well.
> 
> Web 3.0 Technologies Foundation Baarerstrasse 14 6300 Zug
> 
> Many thank for adding the missing information and providing the invoice.
> 
> Best, Melanie

Thanks Melanie, I have emailed and resubmitted the form. Lmk if there is anything further required. Have a nice Christmas.

semuelle commented:
Hey @livetreetech. Sorry, another thing: the entity and address on the invoice doesn't match the ones in the application. Could you update either one to match the other?

livetreetech commented:
Apologies @semuelle  I missed that, I have updated the original application, however, the commit do not seem to be reflecting on the [PR 1818 ](https://github.com/w3f/Grants-Program/pull/1818). @takahser [mentioned](https://github.com/w3f/Grants-Program/pull/1818#issuecomment-1849203081) it might be better to create another PR so, I've done so [here PR 2164](https://github.com/w3f/Grants-Program/pull/2164). For ease I have also attached it [decentral_ml.md](https://github.com/w3f/Grant-Milestone-Delivery/files/13759041/decentral_ml.md) to this comment as well if you wish to simply replace 1818 on your side.  Thanks for the help and have a good festive break.

semuelle commented:
> the commit does not seem to be reflecting on the [PR 1818](https://github.com/w3f/Grants-Program/pull/1818)

Hey @livetreetech. I just left a comment on the open PR. Already merged PRs don't update, so creating a new PR is the right move. Yours contains two files, though. Other than that, it looks good.

livetreetech commented:
Hi @semuelle 
I added the additional file as previous commits did not reflect. Apologies, Git can be frustrating, anyhow, I believe I have now deleted the additional file which can be seen [here](https://github.com/w3f/Grants-Program/pull/2164/commits/675a96f438c22c969b7c90bc2c107fa348473507) Have a great start to the new year. If you could process this as soon as possible it would be appreciated.

livetreetech commented:
Hi @semuelle, thanks for the message, the invoice seems to have the address, I have emailed it to you: s***@web3.foundation let me know, if there is something else required.  Best wishes, A-
